src/ASM/FiniteElement.C

```diff
@@ -173,3 +173,71 @@ bool MxFiniteElement::Hessian (Matrix3D& Hess, const Matrix& Jac,
 
   return ok;
 }
+
+
+void FiniteElement::piolaMapping (const Matrix& J, const Matrix& Ji,
+                                  const Matrices& dNdu, const Matrix3D& H)
+{
+  this->piolaBasis(J);
+  this->piolaGradient(J, Ji, dNdu, H);
+}
+
+
+void FiniteElement::piolaBasis (const Matrix& J)
+{
+  size_t np = 0, dim = J.rows();
+  for (size_t b = 1; b <= dim; ++b)
+    np += this->basis(b).size();
+
+  Matrix Ntmp(dim,np);
+  size_t k = 1;
+  for (size_t b = 1; b <= dim; ++b)
+    for (size_t i = 1; i <= this->basis(b).size(); ++i, ++k)
+      Ntmp(b,k) = this->basis(b)(i);
+
+  P.multiply(J,Ntmp,false,false,false,1.0/detJxW);
+}
+
+
+void FiniteElement::piolaGradient (const Matrix& J, const Matrix& Ji,
+                                   const Matrices& dNdu, const Matrix3D& H)
+{
+  size_t np = 0, dim = J.rows();
+  for (size_t b = 1; b <= dim; ++b)
+    np += this->basis(b).size();
+
+  dPdX.resize(dim*dim,np);
+
+  Matrices dJdX;
+  Vector dDetdX;
+  utl::JacobianGradient(Ji,H,dJdX);
+  utl::detJacGradient(J,Ji,H,dDetdX);
+
+  std::vector<Matrix> Ptmp(dim);
+  for (size_t i = 0; i < dim; ++i)
+  {
+    Ptmp[i] = dJdX[i];
+    Ptmp[i] *= 1.0 / detJxW;
+    Ptmp[i].add(J, -dDetdX[i] / (detJxW * detJxW));
```

Comment:
in the current use i setup the piola stuff before the multiplication with weights and parametric volume/area. but yes, i guess it could be a future bomb. i changed it as it was inconsistent in your refactor (piolabasis used fe.detJxW, the others didn't so i unified them but maybe i should have unified the other way).

Reply:
Just for the record, it may very well have been inconsistent before refactor..missed a 'noticed it was' there.